File: Sources/AINotebookCore/StorePath.swift
import Foundation

/// Where the SQLite database lives. Either an on-disk file URL or an
/// in-memory marker for tests. Pulled out so production code resolves the
/// Application Support path while tests inject a fully in-memory store.
public struct StorePath: Sendable {
    public let fileURL: URL?

    public var isInMemory: Bool { fileURL == nil }

    public static let inMemory = StorePath(fileURL: nil)

    public init(fileURL: URL?) {
        self.fileURL = fileURL
    }

    /// Returns `~/Library/Application Support/AINotebook/db.sqlite`,
    /// creating the parent directory on demand.
    public static func production(
        fileManager: FileManager = .default
    ) throws -> StorePath {
        let appSupport = try fileManager.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let container = appSupport.appendingPathComponent("AINotebook", isDirectory: true)
        try fileManager.createDirectory(
            at: container,
            withIntermediateDirectories: true
        )
        return StorePath(fileURL: container.appendingPathComponent("db.sqlite"))
    }
}

File: windows/src/AINotebook.Core/Storage/StorePath.cs
namespace AINotebook.Core.Storage;

/// <summary>
/// Where the SQLite database lives. Either an on-disk file path or the
/// in-memory marker (FilePath == null) for tests.
/// </summary>
public sealed class StorePath
{
    public string? FilePath { get; }
    public bool IsInMemory => FilePath is null;

    public StorePath(string? filePath) => FilePath = filePath;

    public static StorePath InMemory => new(null);

    /// <summary>
    /// %APPDATA%\AINotebook\db.sqlite, creating the parent directory on demand.
    /// </summary>
    public static StorePath Production()
    {
        var appData = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
        var container = Path.Combine(appData, "AINotebook");
        Directory.CreateDirectory(container);
        return new StorePath(Path.Combine(container, "db.sqlite"));
    }
}
